homeassistant/components/dlna_dms/config_flow.py

```diff
@@ -0,0 +1,177 @@
+        self._discoveries: dict[str, ssdp.SsdpServiceInfo] = {}
+        self._location: str | None = None
+        self._usn: str | None = None
+        self._name: str | None = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initialized by the user by listing unconfigured devices."""
+        LOGGER.debug("async_step_user: user_input: %s", user_input)
+
+        if user_input is not None and (host := user_input.get(CONF_HOST)):
+            # User has chosen a device
+            discovery = self._discoveries[host]
+            await self._async_parse_discovery(discovery)
+            return self._create_entry()
+
+        if not (discoveries := await self._async_get_discoveries()):
+            # Nothing found, abort configuration
+            return self.async_abort(reason="no_devices_found")
+
+        self._discoveries = {
+            cast(str, urlparse(discovery.ssdp_location).hostname): discovery
+            for discovery in discoveries
+        }
+
+        discovery_choices = {
+            host: f"{discovery.upnp.get(ssdp.ATTR_UPNP_FRIENDLY_NAME)} ({host})"
+            for host, discovery in self._discoveries.items()
+        }
+        data_schema = vol.Schema({vol.Optional(CONF_HOST): vol.In(discovery_choices)})
+        return self.async_show_form(step_id="user", data_schema=data_schema)
+
+    async def async_step_ssdp(self, discovery_info: ssdp.SsdpServiceInfo) -> FlowResult:
+        """Handle a flow initialized by SSDP discovery."""
+        LOGGER.debug("async_step_ssdp: discovery_info %s", pformat(discovery_info))
+
+        await self._async_parse_discovery(discovery_info)
+
+        # Abort if the device doesn't support all services required for a DmsDevice.
+        # Use the discovery_info instead of DmsDevice.is_profile_device to avoid
+        # contacting the device again.
+        discovery_service_list = discovery_info.upnp.get(ssdp.ATTR_UPNP_SERVICE_LIST)
+        if not discovery_service_list:
+            return self.async_abort(reason="not_dms")
+        discovery_service_ids = {
+            service.get("serviceId")
+            for service in discovery_service_list.get("service") or []
+        }
+        if not DmsDevice.SERVICE_IDS.issubset(discovery_service_ids):
+            return self.async_abort(reason="not_dms")
+
+        # Abort if another config entry has the same location, in case the
+        # device doesn't have a static and unique UDN (breaking the UPnP spec).
+        self._async_abort_entries_match({CONF_URL: self._location})
+
+        self.context["title_placeholders"] = {"name": self._name}
+
+        return await self.async_step_confirm()
+
+    async def async_step_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Allow the user to confirm adding the device."""
+        LOGGER.debug("async_step_confirm: %s", user_input)
```

Comment:
The user input here is not relevant as it's just a confirm step. I'd remove this log.